### PR TITLE
Add `amdaie.flow` for packet flow lowering

### PR DESCRIPTION
**Pull request by jtuyls**
This PR introduces the `amdaie.flow` operation (in consistency with `aie.flow`). This operation represents connections between ports/channels on tiles in the array and is lowered to `aie.flow` for routing. This idea is to lower `amdaie.connection` to `amdaie.flow` + DMA operations, however this PR goes halfway by creating `amdaie.flow` operations, while still keeping a reference inside `amdaie.connection` as the latter is still needed inside `LowerToAIE` for lowering to the AIE dialect DMA operations. This will be cleaned up in the future, but for now, this enables me to start adding routes for control packets in conjunction with data routes for https://github.com/nod-ai/iree-amd-aie/issues/714.

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEOps.cpp [hunk 1]
```diff
@@ -469,6 +470,22 @@ ConnectionOp::getNpuCircularDmaCpyNdUser() {
   return npuDmaUsers[0];
 }
 
+AMDAIE::FlowOp ConnectionOp::getFlowOp() {
+  return dyn_cast_if_present<AMDAIE::FlowOp>(getFlow().getDefiningOp());
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEOps.cpp:474]: Would returning `std::optinal<AMDAIE::FlowOp>` be better?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEOps.td [hunk 2]
```diff
@@ -91,6 +91,42 @@ def AMDAIE_EndOp: AMDAIE_Op<"end", [Terminator]> {
   let assemblyFormat = [{ attr-dict }];
 }
 
+def AMDAIE_FlowOp: AMDAIE_Op<"flow", [AttrSizedOperandSegments]>, 
+    Results<(outs Index)> {
+  let summary = "The data connection between a set of source and target "
+                "channels.";
+  let description = [{
+    This operation represents a connection between source and target channels. 
+    This is used describe an logical data routing configuration between
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEOps.td:100]:
```suggestion
    This is used to describe a logical data routing configuration between 
```

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEOps.td [hunk 3]
```diff
@@ -91,6 +91,42 @@ def AMDAIE_EndOp: AMDAIE_Op<"end", [Terminator]> {
...
 
+def AMDAIE_FlowOp: AMDAIE_Op<"flow", [AttrSizedOperandSegments]>, 
+    Results<(outs Index)> {
+  let summary = "The data connection between a set of source and target "
+                "channels.";
+  let description = [{
+    This operation represents a connection between source and target channels. 
+    This is used describe an logical data routing configuration between 
+    channels, to be solved by the router for actual stream switch 
+    configurations that implement it. The multiple sources and targets can
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEOps.td:102]:
```suggestion
    configurations that implements it. The multiple sources and targets can
```

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignChannels.cpp [hunk 4]
```diff
@@ -45,20 +45,20 @@ LogicalResult assignChannels(AMDAIE::WorkgroupOp workgroupOp) {
...
       auto channelOp = rewriter.create<AMDAIE::ChannelOp>(
-          rewriter.getUnknownLoc(), tile, channel);
+          rewriter.getUnknownLoc(), tile, channel, StrmSwPortType::DMA);
       targetChannels.push_back(channelOp.getResult());
     }
     rewriter.replaceOpWithNewOp<AMDAIE::ConnectionOp>(
         connectionOp, connectionOp.getTarget(), targetChannels,
         connectionOp.getSource(), sourceChannels,
-        connectionOp.getConnectionTypeAttr());
+        connectionOp.getConnectionTypeAttr(), nullptr);
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignChannels.cpp:61]: Better to describe via `/*argumentName=*/nullptr` here and elsewhere.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignPacketIds.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+      AMDAIE::getDeviceModel(maybeDevice.value());
+  int pktFlowIndex{0};
+  WalkResult res = parentOp->walk([&](AMDAIE::FlowOp flowOp) {
+    if (pktFlowIndex > deviceModel.getPacketIdMaxIdx()) {
+      flowOp.emitOpError() << "ran out of packet IDs to assign";
+      return WalkResult::interrupt();
+    }
+    rewriter.setInsertionPoint(flowOp);
+    auto ui8ty =
+        IntegerType::get(rewriter.getContext(), 8, IntegerType::Unsigned);
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignPacketIds.cpp:49]: Pull this out of the walk.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp [hunk 6]
```diff
@@ -842,6 +831,38 @@ LogicalResult AIEDeviceBuilder::connectionToAIE(
...
+/// Convert the `amdaie.flow` ops into `aie.flow` ops.
+LogicalResult AIEDeviceBuilder::flowToAIE(AMDAIE::FlowOp flowOp,
+                                          Block *deviceBlock) {
+  LLVM_DEBUG(llvm::dbgs() << "Convert [AMDAIE::ConnectionOp]\n");
+  rewriter.setInsertionPointToEnd(deviceBlock);
+  SmallVector<AMDAIE::ChannelOp> producerChannels;
+  SmallVector<AMDAIE::ChannelOp> consumerChannels;
+  for (Value producerChannel : flowOp.getSources()) {
+    auto channelOp =
+        dyn_cast<AMDAIE::ChannelOp>(producerChannel.getDefiningOp());
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELowerToAIE.cpp:843]:
```suggestion
        dyn_cast_if_present<AMDAIE::ChannelOp>(producerChannel.getDefiningOp());
```
Here and below.

**Review by Abhishek-Varma — CHANGES_REQUESTED** [review]

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignPacketIds.cpp (hunk 5 above)
Comment (makslevental) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignPacketIds.cpp:49]: Also please deposit $5 in the `auto` jar (and please remove `auto`)

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEOps.cpp (hunk 1 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/IR/AMDAIEOps.cpp]: Yeah, that would be more clear. Updated it.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignChannels.cpp (hunk 4 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignChannels.cpp]: Done

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignConnectionTypes.cpp [hunk 7]
```diff
@@ -43,7 +43,7 @@ void AMDAIEAssignConnectionTypesPass::runOnOperation() {
     rewriter.replaceOpWithNewOp<AMDAIE::ConnectionOp>(
         connectionOp, connectionOp.getTarget(),
         connectionOp.getTargetChannels(), connectionOp.getSource(),
-        connectionOp.getSourceChannels(), connectionTypeAttr);
+        connectionOp.getSourceChannels(), connectionTypeAttr, nullptr);
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignConnectionTypes.cpp:46]: `/*argumentName=*/`

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConnectionToFlow.cpp [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  // future.
+  WalkResult res = parentOp->walk([&](AMDAIE::ConnectionOp connectionOp) {
+    rewriter.setInsertionPoint(connectionOp);
+    std::optional<AMDAIE::ConnectionType> connectionType =
+        connectionOp.getConnectionType();
+    bool isPacketFlow = connectionType && connectionType.value() ==
+                                              AMDAIE::ConnectionType::Packet;
+    auto flowOp = rewriter.create<AMDAIE::FlowOp>(
+        rewriter.getUnknownLoc(), connectionOp.getSourceChannels(),
+        connectionOp.getTargetChannels(), isPacketFlow, nullptr);
```
Comment (Abhishek-Varma) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConnectionToFlow.cpp:40]: Likewise.

**Review by Abhishek-Varma — APPROVED** [review]
A few more nits to address before merging. LGTM otherwise.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignConnectionTypes.cpp (hunk 7 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignConnectionTypes.cpp]: Done

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConnectionToFlow.cpp (hunk 8 above)
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEConnectionToFlow.cpp]: Done

**Pull request merged**